api/utilities.py

```diff
@@ -137,3 +147,159 @@ async def add_parameter_name_and_datetime(request, parameter_name: str | None, d
+    return [getattr(i.combo, field) for i in response.groups]
+
+
+def filter_observations_for_z(observations, z):
+    if z:
+        z_values = get_z_levels_or_range(z)
+        if len(z.split("/")) == 2:
+            z_min, z_max = z_values[0], z_values[1]
+            observations = [
+                obs
+                for obs in observations
+                if is_float(obs.ts_mdata.level) and z_min <= float(obs.ts_mdata.level) <= z_max
+            ]
+        else:
+            observations = [
+                obs for obs in observations if is_float(obs.ts_mdata.level) and float(obs.ts_mdata.level) in z_values
+            ]
+    return observations
+
+
+def numeric_sort_key(value: str) -> float:
+    """
+    Converts a string to a float for comparison, returns infinity if the string is not convertible.
+    """
+    try:
+        return float(value)
+    except ValueError:
+        return float("inf")
+
+
+def iso_8601_duration_to_seconds_sort_key(duration: str) -> int:
+    """
+    Converts an ISO 8601 duration string to seconds for comparison.
+    Can compare day, hour, minute, and second values or a combination
+    """
+    # Regular expression to match the ISO 8601 duration string.
+    # Splits the expression into days, hours, minutes, and seconds
+    # and calculates total_seconds for comparison.
+    pattern = re.compile(r"P(?:(\d+)D)?T(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?")
+    match = pattern.match(duration)
+    if not match:
+        # If the duration string is not in the correct format, return the maximum integer value
+        return sys.maxsize
+
+    days = int(match.group(1) or 0)
+    hours = int(match.group(2) or 0)
+    minutes = int(match.group(3) or 0)
+    seconds = int(match.group(4) or 0)
+
+    total_seconds = days * 86400 + hours * 3600 + minutes * 60 + seconds
+    return total_seconds
+
+
+async def get_iso_8601_range(start: str, end: str) -> list[str] | None:
+    """
+    Returns a list of ISO 8601 durations between the start and end values.
+    TODO: Add support for closest match if the start or end value is valid, but not found?
+    """
+
+    if not start or not end:
+        raise HTTPException(status_code=400, detail=f"Invalid ISO 8601 period: {start} / {end}")
+
+    try:
+        periods = sorted(await get_unique_values_for_metadata("period"), key=iso_8601_duration_to_seconds_sort_key)
+
+        if start != "..":
+            start_index = periods.index(start)
+        else:
+            start_index = 0
+
+        if end != "..":
+            end_index = periods.index(end)
+        else:
+            end_index = len(periods) - 1
+
+        if start_index > end_index:
+            raise HTTPException(status_code=400, detail=f"Invalid ISO 8601 range: {start} > {end}")
+
+    except ValueError as err:
+        raise HTTPException(status_code=400, detail=f"Invalid ISO 8601 range: {err} of possible period values")
+
+    return periods[start_index : end_index + 1]  # noqa: E203
```

Comment:
Better range support in: https://github.com/EURODEO/e-soh/pull/146/commits/f1196f886182413fc635e5b621296691c5820e29

Decided to make a separate function for transforming durations to seconds, this function is in line with what @Teddy-1000 did in the upcoming ingest change. The sorting functions purpose is to sort, so I kept it as such. It will be obsolete anyways after storing periods as seconds in the database.

Reply:
I am confused... the `iso_8601_duration_to_seconds_sort_key` function does exactly the same as `iso_8601_duration_to_seconds`, so why keep it?

This looks pretty clear to me:
`sorted(await get_unique_values_for_metadata("period"), key=iso_8601_duration_to_seconds)`

If you really want to keep the `iso_8601_duration_to_seconds_sort_key`, just call the other function from there.
Besides the maintenance, having two functions which try to do the same thing in a different way might give some really confusing results...

Or am I missing something?

Reply:
Almost exactly the same. Other one is used to `sort` durations inside the system and it will not hinder the system unusable and raise an error if the system for some reason contains data which cannot be sorted, it handles it "gracefully" and just pushes them at the end of the list. We can get rid of it after everything is stored as integers, since they are always sortable.

The other one is used to transform durations to seconds, but also `validate` user queries and will raise an error if it finds a duration which cannot be transformed to seconds. 

Using the former for example with sorting the periods for `/collections` with unsortable data would result it raising an error and I don't think that is how custom sorting algorithm should work.